pkg/blockbuilder/tsdb.go

```diff
@@ -0,0 +1,398 @@
+		if err := db.compactEverything(ctx); err != nil {
+			return numBlocks, err
+		}
+
+		var blockNames []string
+		dbDir := db.db.Dir()
+		for _, b := range db.db.Blocks() {
+			blockNames = append(blockNames, b.Meta().ULID.String())
+		}
+		numBlocks += len(blockNames)
+
+		if err := db.Close(); err != nil {
+			return numBlocks, err
+		}
+
+		delete(b.tsdbs, tenant)
+
+		// Make a copy of userID because 'tenant' changes in the next iteration.
+		userID := tenant.id
+		eg.Go(func() error {
+			uploader := blockUploaderForUser(ctx, userID)
+			for _, bn := range blockNames {
+				if err := uploader(filepath.Join(dbDir, bn)); err != nil {
+					return err
+				}
+			}
+
+			// Clear the DB from the disk. Don't need it anymore.
+			return os.RemoveAll(dbDir)
+		})
+	}
+	err := eg.Wait()
+
+	// Clear the map so that it can be released from the memory. Not setting to nil in case
+	// we want to reuse the TSDBBuilder.
+	b.tsdbs = make(map[tsdbTenant]*userTSDB)
+	return numBlocks, err
+}
+
+func (b *TSDBBuilder) Close() error {
+	b.tsdbsMu.Lock()
+	defer b.tsdbsMu.Unlock()
+
+	for userID, db := range b.tsdbs {
+		dbDir := db.db.Dir()
+
+		if err := db.Close(); err != nil {
+			return err
+		}
+		delete(b.tsdbs, userID)
+
+		// Remove any artifacts of this TSDB. We don't need them anymore.
+		if err := os.RemoveAll(dbDir); err != nil {
+			return err
+		}
+	}
+
+	// Clear the map so that it can be released from the memory. Not setting to nil in case
+	// we want to reuse the TSDBBuilder.
+	b.tsdbs = make(map[tsdbTenant]*userTSDB)
+	return nil
+}
+
+type extendedAppender interface {
+	storage.Appender
+	storage.GetRef
+}
+
+type userTSDB struct {
+	db     *tsdb.DB
+	userID string
+}
+
+// BlocksUploader interface is used to have an easy way to mock it in tests.
+type BlocksUploader interface {
+	Sync(ctx context.Context) (uploaded int, err error)
+}
+
+func (u *userTSDB) Head() *tsdb.Head {
+	return u.db.Head()
+}
+
+func (u *userTSDB) Close() error {
+	return u.db.Close()
+}
+
+func (u *userTSDB) Appender(ctx context.Context) storage.Appender {
+	return u.db.Appender(ctx)
+}
+
+func (u *userTSDB) compactEverything(ctx context.Context) error {
+	blockRange := 2 * time.Hour.Milliseconds()
+
+	// Compact the in-order data.
+	mint, maxt := u.Head().MinTime(), u.Head().MaxTime()
+	mint = (mint / blockRange) * blockRange
+	for blockMint := mint; blockMint <= maxt; blockMint += blockRange {
+		blockMaxt := blockMint + blockRange - 1
+		rh := tsdb.NewRangeHead(u.Head(), blockMint, blockMaxt)
+		if err := u.db.CompactHeadWithoutTruncation(rh); err != nil {
+			return err
+		}
+	}
+
+	// Compact the out-of-order data.
+	if err := u.db.CompactOOOHead(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *userTSDB) PreCreation(_ labels.Labels) error                     { return nil }
+func (u *userTSDB) PostCreation(_ labels.Labels)                          {}
+func (u *userTSDB) PostDeletion(_ map[chunks.HeadSeriesRef]labels.Labels) {}
```

Comment:
It's still to be built. In general, all limits are still to be built because all of them have a fundamental problem of how to apply when they change, i.e. ingesters will apply the current limits to current data, but block builder will be applying current limits to past data, so its application will differ from ingesters if the config was changed.

Reply:
Ok this is quite concerning to me. I would like to see this addressed asap.